xc/common/utils/prjxray_edge_library.py

```diff
@@ -1149,6 +1149,91 @@ def make_connection(
         yield edge
 
 
+def kill_track(conn, roi, synth_tracks):
+    """ Checks tracks with connections outside of the ROI.
+
+    Iterates over all alive tracks and removes the alive flag if
+    the track can lead to connections outside of the ROI.
+
+    Args:
+        conn (sqlite3.Connection): Connection database
+        roi: prjxray ROI object
+        synth_tracks : tracks related to synth tile connections (cannot be removed)
+
+    """
+
+    killed_tracks = set()
+    c = conn.cursor()
+    write_cur = conn.cursor()
+
+    for (track_pkey, ) in c.execute("""
+SELECT
+  pkey
+FROM
+  track
+WHERE
+  alive;"""):
+
+        write_cur.execute(
+            """
+SELECT vcc_track_pkey, gnd_track_pkey FROM constant_sources;
+        """
+        )
+
+        vcc_track_pkey, gnd_track_pkey = write_cur.fetchone()
+
+        if track_pkey in killed_tracks:
+            continue
+
+        if track_pkey in synth_tracks:
+            continue
+
+        if track_pkey in [vcc_track_pkey, gnd_track_pkey]:
+            continue
+
+        write_cur.execute(
+            """
+SELECT track_pkey FROM graph_node WHERE pkey = (
+  SELECT dest_graph_node_pkey FROM graph_edge WHERE src_graph_node_pkey = (
+    SELECT pkey FROM graph_node WHERE track_pkey = ?));""", (track_pkey, )
+        )
+
+        (dest_track_pkey, ) = write_cur.fetchone()
+        write_cur.execute(
+            """
+SELECT grid_x, grid_y FROM phy_tile WHERE pkey = (
+SELECT canon_phy_tile_pkey FROM track WHERE pkey = ?);""", (dest_track_pkey, )
+        )
+
+        tile_info = write_cur.fetchone()
+
+        if tile_info:
```

Comment:
We should be removing edges, not nodes.  Just because a node has an edge that is outside of the ROI doesn't mean it it should be killed.

If an **edge** (e.g. pip) exists outside of the ROI it should be removed, and if an **edge** connects **to** an output only node, or an **edge** connects **from** an input only node it should be removed.  The alive check will then kill any nodes that are unreachable or are dead ends.

Reply:
As far as I have seen, the edge to and from the IO only nodes is already taken care of.

The remaining problem would be removing the edges existing outside of the ROI. I thought about this, and it probably does not completely solve the issue of producing shorted connections.

For instance, an LH12 wire that has three endpoints can have two of the endpoints within the ROI and the last one outside of the ROI. This means that even though the edges present outside of the ROI are removed, the LH12 node will still be present as it has valid graph edges, therefore needs to be marked as alive, with the same potential of generating shorted wire.

Reply:
> For instance, an LH12 wire that has three endpoints can have two of the endpoints within the ROI and the last one outside of the ROI. This means that even though the edges present outside of the ROI are removed, the LH12 node will still be present as it has valid graph edges, therefore needs to be marked as alive, with the same potential of generating shorted wire.

I've already talked about this.  In the case of the LH12, yes there are multiple endpoints.  However the entire node should be marked as "input only" (e.g. this is connected in the ROI to a source), or "output only" (e.g. this is connected in the ROI to a sink).  An "input only" node should only have edges that come from the node (except for the incoming edge from the SYN-IN).  An "output only" node should only have edges that come to the node (except for the outcoming edge to the SYN-OUT).  There can be no shorts if this is actually done.

Reply:
I am still kind of confused about this.

The problematic nodes that cannot be used are not related to SYN-IN, SYN-OUT or constant sources.
In `mark_track_liveness`, if a track is present in the "input_only_nodes" or "output_only_nodes" it will get added to the alive tracks, so adding the problematic nodes to the input only set or output only set won't get that node removed.

What I could do, is to have the edges creation process to avoid adding edges if the node is an "input only" node and the edge connects from that node (or vice-versa). The problem though is that I do not understand how to detect which are the problematic nodes prior to the creation of the graph edges.

Moreover, I do not yet understand why we can't just kill a track that is not related to synth nodes or constant source nodes and has edges outside the ROI, as it happens in the current implementation. After all, the node will still be killed if it has no output edges or no input edges during `mark_track_liveness` and it was marked as input- or output-only node.

Reply:
Let's take a step back.  I thought the original bug was because a node that is part of the ROI harness was being include in the routing graph, and VPR used it for general routing.  All nodes that are part of the ROI harness should be modified such that the routing graph doesn't allow them to be used in a way that would break the ROI harness.  In the original example, I believe that the failure was a short was created because a LH12 wire had two drivers, one from VPR and one from the ROI harness?  If the relevant node was marked as either input_only or output_only, the relevant edges for that node would be excluded to prevent a misuse/short.  Am I missing something in that description?

Reply:
> I thought the original bug was because a node that is part of the ROI harness was being included in the routing graph, and VPR used it for general routing

As far as I understood yes, this is the main issue. The RR graph includes nodes that are being used in the harness (in this  particular situation, the node is `INT_R_X19Y135/LH12`, which can span out of the ROI).

>  I believe that the failure was a short was created because a LH12 wire had two drivers, one from VPR and one from the ROI harness?

Correct, the above-mentioned wire is used by VPR as a "bridge" to connect two nodes in the `INT_R_X19Y135` tile, while, the other end of the wire is used by the ROI harness and is connected to drive a SYN-IN.

>  If the relevant node was marked as either input_only or output_only, the relevant edges for that node would be excluded to prevent a misuse/short

I think that this can be done, but we would require information on which nodes are used in the ROI harness. I see that we can get this information from the design.json of the harness, which has a field called `wires_outside_roi`, which could potentially be used to get the wires that need to be marked as input_only or output_only.
The problem here is that the same wire, can be named differently in the `design.json`. To be more precise, the `INT_R_X19Y135/LH12` is not present in the json, but instead, `INT_R_X25Y135/LH6` is. Now, the latter is not a node, but a pip in the INT tile. The two though are the exact same wire.

In conclusion, I believe this issue could be "easily" solved by using the information on the wires outside the roi present in the json, but we would need to provide correct naming of the nodes used in the harness.

Reply:
> I see that we can get this information from the design.json of the harness, which has a field called wires_outside_roi, which could potentially be used to get the wires that need to be marked as input_only or output_only.

Yes, and I believe it was intended to be used for that purpose.  The connection database can translate the wire -> node.